### DISCUSSION TITLE
Multipart handler not respecting RequestBodyLimitLayer adjustment

### DISCUSSION THREAD

**Discussion by sneakycrow** [tokio-rs/axum#1666] [Q&A] [answered]
I have a web service that deploys an SPA as it's primary route and a multipart upload route. I'm new to axum and I've put this together mostly from copying the examples. Even though I'm adjusting the `RequestBodyLimitLayer`, the upload route still gets a `MultipartError` for length limit exceeded on anything above 2 MB

```rust
use std::net::SocketAddr;

use axum::extract::Multipart;
use axum::{routing::post, Router};
use axum_extra::routing::SpaRouter;
use tower_http::limit::RequestBodyLimitLayer;

const CONTENT_LENGTH_LIMIT: usize = 20 * 1024 * 1024;

#[tokio::main]
async fn main() {
    tracing_subscriber::fmt::init();
    let index = SpaRouter::new("/", "web/dist").index_file("index.html");
    let app: Router = Router::new()
        .merge(index)
        .layer(RequestBodyLimitLayer::new(CONTENT_LENGTH_LIMIT))
        .route("/upload", post(upload));

    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    tracing::debug!("listening on http://{}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn upload(mut multipart: Multipart) {
    while let Some(mut field) = multipart.next_field().await.unwrap() {
        let name = field.name().unwrap().to_string();
        let data = field.bytes().await.unwrap();

        println!("Length of `{}` is {} bytes", name, data.len());
    }
}
```

Here's the error I receive when trying to upload a 2.1 MB zip:
`thread 'tokio-runtime-worker' panicked at 'called Result::unwrap() on an Err value: MultipartError { source: failed to read stream: failed to read stream: length limit exceeded }', api/src/main.rs:31:40`

And here's the http request itself if that's helpful

```http
POST http://localhost:3000/upload
Content-Type: multipart/form-data; boundary=boundary

--boundary
Content-Disposition: form-data; name="file"; filename="file.zip"

< ./file.zip
```

How can I adjust the code above so that the upload route will accept files below my custom limit `CONTENT_LENGTH_LIMIT`?

**Comment by davidpdrsn** [comment]
You need to apply the layer after adding the upload route, not before.

**Reply by sneakycrow** [reply]
> You need to apply the layer after adding the upload route, not before.

Do you mean like this? I tried changing it to the below and I still get the same `MultipartError`.

```rust
    let app: Router = Router::new()
        .merge(index)
        .route("/upload", post(upload))
        .layer(RequestBodyLimitLayer::new(CONTENT_LENGTH_LIMIT));
```

**Reply by davidpdrsn** [reply]
Yep thats what I meant. Not sure why its not working. I'm on holiday now but can look into it next week.

**Reply by sneakycrow** [reply]
Thanks @davidpdrsn, appreciate the quick response. I'll keep an eye out!

**Comment by sneakycrow** [comment, marked as answer]
I was able to get it to respect the `RequestBodyLimitLayer` by disabling the `DefaultBodyLimit` in the layer before, like this:

```rust
    let app: Router = Router::new()
        .merge(index)
        .route("/upload", post(upload))
        .layer(DefaultBodyLimit::disable())
        .layer(RequestBodyLimitLayer::new(CONTENT_LENGTH_LIMIT));
```

Is this intended functionality? I didn't see this being a requirement in any examples.

Edit:

Per @jplatte's suggestion, this works too and is much cleaner:

```rust
    let app: Router = Router::new()
        .merge(index)
        .route("/upload", post(upload))
        .layer(DefaultBodyLimit::max(CONTENT_LENGTH_LIMIT));
```

**Reply by davidpdrsn** [reply]
Yes that's expected. It's documented here https://docs.rs/axum/latest/axum/extract/struct.DefaultBodyLimit.html

**Reply by jplatte** [reply]
Shouldn't the single layer work if it is `DefaultBodyLimit::max(CONTENT_LENGTH_LIMIT)` instead of `RequestBodyLimitLayer::new(CONTENT_LENGTH_LIMIT)`?

**Reply by sneakycrow** [reply]
> Shouldn't the single layer work if it is `DefaultBodyLimit::max(CONTENT_LENGTH_LIMIT)` instead of `RequestBodyLimitLayer::new(CONTENT_LENGTH_LIMIT)`?

This does work and is much cleaner. Thanks for the response, I'll update my original post with a correct answer.
